.github/workflows/README.md

```diff
@@ -14,6 +14,87 @@
+
+## Token Permissions
+
+You should explicitly define the GitHub Actions token scopes needed by your job. For most jobs, this will be `actions: write` (needed for comment triggering support) and `read` permissions for all other options. See https://github.com/apache/beam/blob/907c5110163b0efe52e9e12127fd013c7fc455d7/.github/workflows/beam_PreCommit_Go.yml#L16 for an example (you can copy and paste this into your workflow).
+
+## Concurrency Groups
+
+Concurrency groups are a way of making sure that no more than one Actions run is running per job/group at any given time (previous ones can be cancelled). To reduce resource usage, you should define the following concurrency group:
+
+```
+concurrency:
+  group: '${{ github.workflow }} @ ${{ github.event.pull_request.head.label || github.sha || github.head_ref || github.ref }}-${{ github.event.sender.login }}-${{ github.event.schedule }}'
+  cancel-in-progress: true
+```
+
+this defines the following groups:
+
+1) Each workflow will represent a different set of groups
+2) Within each workflow, each PR will represent a single group
+3) Within each non-PR run for a workflow, each commit will represent a set of groups
+4) Within each commit, each push event, schedule event, and manual run event will represent a set of groups
+
+## Comment Triggering Support
+
+In order to make it easier for non-committers to interact with workflows, workflows should implement comment triggering support. This requires 3 additional components beyond the ones mentioned above:
+
+1) Each job should be gated on an if condition that maps to that workflow's comment trigger (example: https://github.com/apache/beam/blob/907c5110163b0efe52e9e12127fd013c7fc455d7/.github/workflows/beam_PreCommit_Go.yml#L38)
+2) Each job should have the rerun action immediately after its checkout step. This should be gated on the comment trigger (example: https://github.com/apache/beam/blob/907c5110163b0efe52e9e12127fd013c7fc455d7/.github/workflows/beam_PreCommit_Go.yml#L49)
+3) Each job should have a descriptive name that includes the comment trigger (example: https://github.com/apache/beam/blob/ba8fc935222aeb070668fbafd588bc58e7a21289/.github/workflows/beam_PreCommit_CommunityMetrics.yml#L48)
+
+# Testing new workflows or workflow updates
+
+## Testing New Workflows
+
+New workflows are not added to the [UI on the Actions tab](https://github.com/apache/beam/actions) until they are merged into the repo's main branch (master for Beam).
+To test new workflows, we recommend the following pattern:
+
+1) Fork the Beam repo
+2) Add your proposed workflow to the main branch of your fork.
+3) Run the workflow in the [Actions tab](https://github.com/apache/beam/actions) of your fork using the `Run workflow` button
```

Comment:
If I were to add a test that will run in apache-beam-testing project, what kind of credentials would i need to run such test? guessing that uploading any service accounts would also be a security risk, so should we advise to request committers help to trigger test runs or mention that we can only test plumbing of the action but test execution will have to wait until the suite is merged?

Reply:
So we've basically decided to live with that risk for now; it is ok to provide access to limited service accounts when needed to adequately test pipelines. Typically, that only exposes using our compute as an attack vector since pipelines basically just need VMs or Dataflow to run, which is something we can respond to if it becomes an issue. The same set of risks exist on Jenkins today (since precommits can start arbitrary jobs using service account permissions on the boxes, and we decided to not take on the burden of fixing our full test model as part of this migration.

There's basically no way to allow users to run tests that use compute without either:
1) This risk
2) Requiring manual intervention for every precommit

Reply:
Letting GH actions to run Dataflow pipelines seems reasonable. My question is about how to advise non-committers to test their actions workflows on Dataflow. Sounds like a committer would be able to fully test a suite using stored credentials for GH action, if they create a branch on Apache Repo. Non-committers wouldn't be able to create branches on Apache Repo, but they can test against branches created on their own fork; however their own fork would presumably not preconfigured GCP secrets. So likely, their gh action test will start, but dataflow pipeline will fail to launch, and that might be OK. We probably shouldn't advise customers to upload their own secrets to their own gh actions..

Reply:
To make expectations clear we could say smth like: you would only be able to do X and Y but not Z, and if you really want to do Z you'd need Beam's committer help or wait until PR is merged or do XYZ (configure your own credentials, which I am hesitant to recommend).